transform/log_address_index_provider.go

```diff
@@ -0,0 +1,168 @@
+package transform
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"time"
+
+	"github.com/streamingfast/dstore"
+	"github.com/streamingfast/eth-go"
+	"go.uber.org/zap"
+)
+
+const LogAddressIdxShortname string = "logaddr"
+
+// LogAddressIndexProvider responds to queries about LogAddressIndexes
+type LogAddressIndexProvider struct {
+	currentIndex          *LogAddressIndex
+	currentMatchingBlocks []uint64
+	filterAddresses       []eth.Address
+	filterEventSigs       []eth.Hash
+	indexOpsTimeout       time.Duration
+	possibleIndexSizes    []uint64
+	store                 dstore.Store
+}
+
+func NewLogAddressIndexProvider(
+	store dstore.Store,
+	filterAddresses []eth.Address,
+	filterEventSigs []eth.Hash,
+	possibleIndexSizes []uint64,
+) *LogAddressIndexProvider {
+	// todo(froch, 20220207): firm up what these values will be
+	if possibleIndexSizes == nil {
+		possibleIndexSizes = []uint64{100000, 10000, 1000, 100}
+	}
+	if len(filterAddresses) == 0 && len(filterEventSigs) == 0 {
+		return nil
+	}
+
+	return &LogAddressIndexProvider{
+		store:              store,
+		indexOpsTimeout:    15 * time.Second,
+		filterAddresses:    filterAddresses,
+		filterEventSigs:    filterEventSigs,
+		possibleIndexSizes: possibleIndexSizes,
+	}
+}
+
+// WithinRange determines the existence of an index which includes the provided blockNum
+// it also attempts to pre-emptively load the index (read-ahead)
+func (ip *LogAddressIndexProvider) WithinRange(blockNum uint64) bool {
+	r, lowBlockNum, indexSize := ip.findIndexContaining(blockNum)
+	if r == nil {
+		return false
+	}
+	if err := ip.loadIndex(r, lowBlockNum, indexSize); err != nil {
+		zlog.Error("couldn't load index", zap.Error(err))
+		return false
+	}
+	return true
+}
+
+// Matches returns true if the provided blockNum matches entries in the index
+func (ip *LogAddressIndexProvider) Matches(blockNum uint64) (bool, error) {
+	if err := ip.loadRange(blockNum); err != nil {
+		return false, fmt.Errorf("couldn't load range: %s", err)
+	}
+
+	for _, matchingBlock := range ip.currentMatchingBlocks {
+		if blockNum == matchingBlock {
```

Comment:
ideally, this is just a containment check directly in the Roaring bitmap, instead of a loop.

Reply:
https://pkg.go.dev/github.com/RoaringBitmap/roaring/roaring64#Bitmap.Contains

Reply:
hmm, je vois que c'était pour avoir ton NextMatching hein?

Reply:
Il faudrait ça: https://github.com/RoaringBitmap/roaring/issues/55 
On l'implémente? :)

Reply:
`hmm, je vois que c'était pour avoir ton NextMatching hein?` <--- exactement
plusieurs changements sont nécessaires pour utiliser le Contains() et le pattern de l'itérateur.
le NextMatching() va être appelé beaucoup plus souvent que le Matches, itérer dans une liste de 10000 sera pas trop long considérant le temps qu'on sauve à ne pas processer les blocks.

Saving this for next optimization ^^